### PR TITLE
Make `Res::SelfTy` a struct variant and update docs

### PR DESCRIPTION
**Pull request by BoxyUwU**
I found pattern matching on a `(Option<DefId>, Option<(DefId, bool)>)` to not be super readable, additionally the doc comments on the types in a tuple variant aren't visible anywhere at use sites as far as I can tell (using rust analyzer + vscode)

The docs incorrectly assumed that the `DefId` in `Option<(DefId, bool)>` would only ever be for an impl item and I also found the code examples to be somewhat unclear about which `DefId` was being talked about.

r? @lcnr since you reviewed the last PR changing these docs

### CODE REVIEW

File: src/librustdoc/clean/utils.rs [hunk 1]
```diff
@@ -398,10 +398,10 @@ crate fn register_res(cx: &mut DocContext<'_>, res: Res) -> DefId {
             i,
         ) => (i, kind.into()),
         // This is part of a trait definition; document the trait.
-        Res::SelfTy(Some(trait_def_id), _) => (trait_def_id, ItemType::Trait),
+        Res::SelfTy { trait_: Some(trait_def_id), alias_to: _ } => (trait_def_id, ItemType::Trait),
         // This is an inherent impl; it doesn't have its own page.
-        Res::SelfTy(None, Some((impl_def_id, _))) => return impl_def_id,
-        Res::SelfTy(None, None)
+        Res::SelfTy { trait_: None, alias_to: Some((impl_def_id, _)) } => return impl_def_id,
+        Res::SelfTy { trait_: None, alias_to: None }
```
Comment (BoxyUwU) [inline src/librustdoc/clean/utils.rs:404]: I am not familiar with rustdoc so am not sure if this is actually a bug but these comments seem wrong :sweat_smile: 
the `trait_: Some(trait)` case is not _just_ for trait definitions and is also there in `impl Trait for Foo` blocks
the `trait_: None, alias_to: Some(..)` is not _just_ for inherent impls and is also there in struct defs (`struct Foo(Box<Self>)`)

File: compiler/rustc_hir/src/def.rs [hunk 2]
```diff
@@ -266,59 +266,67 @@ pub enum Res<Id = hir::HirId> {
...
+    /// }
+    /// ```
+    // FIXME(generic_const_exprs): Remove this bodge once that feature is stable.
+    SelfTy {
+        /// The trait this `Self` is a generic arg for.
+        trait_: Option<DefId>,
+        /// The item introducing the `Self` type alias. Can be used in the `type_of` query
+        /// to get the underlying type. Additionally whether the `Self` type is disallowed
+        /// from mentioning generics (i.e. when used in an anonymous constant).
+        alias_to: Option<(DefId, bool)>,
```
Comment (camelid) [inline compiler/rustc_hir/src/def.rs:328]: Thanks for naming these fields and improving the docs! I had a lot of trouble understanding this when I wrote the current version of the docs.

Why is `alias_to` an `Option`, though? When would it be `None`?

File: src/librustdoc/clean/utils.rs (hunk 1 above)
Comment (camelid) [reply src/librustdoc/clean/utils.rs:404]: If you can, please go ahead and update these comments so they're correct :)

Comment (lcnr) [reply src/librustdoc/clean/utils.rs:404]: i think it makes sense to add a fixme to potentially change this to
```rust
Res::SelfTy { trait_: Some(trait_def_id), alias_to: None } => (trait_def_id, ItemType::Trait),
```
i think the mention of `impl Send for [u8]` might be caused by this https://doc.rust-lang.org/nightly/std/marker/trait.Send.html#impl-Send-80

but i don't want to add functionality changes to rustdoc in this PR

**Comment by lcnr** [comment]
after adding a fixme to rustdoc, r=me

File: compiler/rustc_hir/src/def.rs (hunk 2 above)
Comment (camelid) [reply compiler/rustc_hir/src/def.rs:328]: Ah, makes sense, thanks :)

Could you add a comment about that?

Comment (camelid) [reply compiler/rustc_hir/src/def.rs:328]: Oops, just realized this was already merged 😅

Comment (BoxyUwU) [reply compiler/rustc_hir/src/def.rs:328]: There's a comment about this if you look at the examples above the `SelfTy` variant

Comment (camelid) [reply compiler/rustc_hir/src/def.rs:328]: Thanks!

**Pull request merged**